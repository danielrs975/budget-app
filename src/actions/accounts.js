/**
 * This script contain all the actions about the
 * accounts information
 * @author Daniel Rodriguez
 */
import uuid from 'uuid';

/**
 * This action add an account to the database
 * @param {*} account all the info about the account
 */
export const addAccount = (account) => ({
	type    : 'ADD_ACCOUNT',
	account
});

export const startAddAccount = (account) => {
	return (dispatch, getState) => {
		dispatch(addAccount({ id: uuid(), ...account }));
		const accounts = getState().accounts;
		localStorage.setItem('accounts', JSON.stringify(accounts));
	};
};

/**
 * This action is set in the state the accounts that are saved
 * in the db
 * @param {*} accounts The accounts to set in the state
 */
export const setAccount = (accounts) => ({
	type     : 'SET_ACCOUNTS',
	accounts
});

export const startSetAccount = () => {
	return (dispatch) => {
		const accounts = JSON.parse(localStorage.getItem('accounts'));
		dispatch(setAccount(accounts || []));
	};
};
