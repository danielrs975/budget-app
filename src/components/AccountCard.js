/**
 * This component contain a short information about an account
 * @author Daniel Rodriguez
 */
import React from 'react';

const AccountCard = ({ name, amount }) => (
	<div className="card-layout">
		<h3>{name}</h3>
		<p>{amount / 100}</p>
	</div>
);

export default AccountCard;
