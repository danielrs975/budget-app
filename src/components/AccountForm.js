import React from 'react';

export default class AccountForm extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			name     : '',
			amount   : '',
			type     : 'bank',
			currency : 'euro',
			error    : undefined
		};
	}
	onNameChange = (e) => {
		const name = e.target.value;
		this.setState(() => ({ name }));
	};
	onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};
	onTypeChange = (e) => {
		const type = e.target.value;
		this.setState(() => ({ type }));
	};
	onCurrencyChange = (e) => {
		const currency = e.target.value;
		this.setState(() => ({ currency }));
	};
	saveAccount = (e) => {
		e.preventDefault();
		if (!this.state.name || !this.state.currency) {
			this.setState(() => ({ error: 'A name for the account must be provided' }));
		} else {
			this.setState(() => ({ error: undefined }));
			this.props.onSubmit({
				name     : this.state.name,
				amount   : parseFloat(this.state.amount) * 100,
				type     : this.state.type,
				currency : this.state.currency
			});
		}
	};
	render() {
		return (
			<form onSubmit={this.saveAccount}>
				{this.state.error && <p>{this.state.error}</p>}
				<input type="text" placeholder="Name of account" value={this.state.name} onChange={this.onNameChange} />
				<input
					type="text"
					placeholder="Initial amount"
					value={this.state.amount}
					onChange={this.onAmountChange}
				/>
				<select placeholder="Type" value={this.state.type} onChange={this.onTypeChange}>
					<option value="bank">Bank</option>
					<option value="cash">Cash</option>
				</select>
				<select placeholder="Currency" value={this.state.currency} onChange={this.onCurrencyChange}>
					<option value="euro">Euro</option>
					<option value="dollar">Dollar</option>
				</select>
				<button>Save account</button>
			</form>
		);
	}
}
