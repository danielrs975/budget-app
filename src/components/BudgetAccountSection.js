/**
 * This component is a container to contains all
 * the accounts and budgets registered
 * @Author: Daniel Rodriguez
 * @Email: danielrs9705@gmail.com
 */
import React from 'react';
import { connect } from 'react-redux';
import AccountsList from './AccountsList';

const BudgetAccountSection = (props) => (
	<div className="content-container">
		<div>
			<h1>Accounts</h1>
			<AccountsList accounts={props.accounts} />
		</div>
	</div>
);

const mapStateToProps = (state) => ({
	accounts : state.accounts
});

export default connect(mapStateToProps)(BudgetAccountSection);
