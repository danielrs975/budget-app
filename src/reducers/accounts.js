/**
 * This is the reducer for the accounts
 * information
 * @author Daniel Rodriguez
 */

export default (state = [], action) => {
	switch (action.type) {
		case 'ADD_ACCOUNT':
			return [ ...state, action.account ];
		default:
			return state;
	}
};
