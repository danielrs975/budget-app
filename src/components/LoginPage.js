import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
	<div className="box-layout">
		<div className="box-layout__box">
			<h1 className="box-layout__title">Budget App</h1>
			<p>Manage your expenses everywhere</p>
			<input type="text" placeholder="E-mail" />
			<input type="password" placeholder="Password" />
			<button className="button">Login</button>
			<p>or</p>
			<button onClick={startLogin} className="button">
				Login with Google
			</button>
		</div>
	</div>
);

const mapDispatchToProps = (dispatch) => ({
	startLogin : () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
