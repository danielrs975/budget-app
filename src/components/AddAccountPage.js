import React from 'react';
import { connect } from 'react-redux';
import AccountForm from './AccountForm';
import { startAddAccount } from '../actions/accounts';

class AddAccountPage extends React.Component {
	onSubmit = (account) => {
		this.props.startAddAccount(account);
		this.props.history.push('/dashboard');
	};
	render() {
		return (
			<div>
				<h1>Add a new account</h1>
				<AccountForm onSubmit={this.onSubmit} />
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({
	startAddAccount : (account) => dispatch(startAddAccount(account))
});

export default connect(undefined, mapDispatchToProps)(AddAccountPage);
