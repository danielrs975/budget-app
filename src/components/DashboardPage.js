import React from 'react';
import BudgetAccountSection from './BudgetAccountSection';

const ExpenseDashboardPage = () => (
	<div>
		<BudgetAccountSection />
	</div>
);

export default ExpenseDashboardPage;
