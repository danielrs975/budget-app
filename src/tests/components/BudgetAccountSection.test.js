import React from 'react';
import { shallow } from 'enzyme';
import { BudgetAccountSection } from '../../components/BudgetAccountSection';
import accounts from '../fixtures/accounts';

test('should BudgetAccountSection component render correcctly', () => {
    const wrapper = shallow(<BudgetAccountSection accounts={accounts}/>);
    expect(wrapper).toMatchSnapshot();
})