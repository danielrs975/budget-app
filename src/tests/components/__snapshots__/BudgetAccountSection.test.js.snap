// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should BudgetAccountSection component render correcctly 1`] = `
<div
  className="content-container"
>
  <div>
    <h1>
      Accounts
    </h1>
    <AccountsList
      accounts={
        Array [
          Object {
            "amount": 1000,
            "currency": "euro",
            "id": "1",
            "name": "Revolut",
            "type": "bank",
          },
          Object {
            "amount": 1500,
            "currency": "euro",
            "id": "2",
            "name": "Societe Generale",
            "type": "bank",
          },
          Object {
            "amount": 5400,
            "currency": "euro",
            "id": "3",
            "name": "Cash",
            "type": "cash",
          },
        ]
      }
    />
  </div>
</div>
`;
